sw/device/lib/crypto/drivers/aes.c

```diff
@@ -67,56 +76,96 @@ aes_error_t aes_begin(aes_params_t params) {
+ */
+static aes_error_t aes_begin(aes_key_t key, const uint32_t iv[4],
+                             bool encrypt) {
+  aes_error_t err = spin_until(AES_STATUS_IDLE_BIT);
+  if (err != kAesOk) {
+    return err;
+  }
+
   uint32_t ctrl_reg = AES_CTRL_SHADOWED_REG_RESVAL;
+  // Set the operation (encrypt or decrypt).
   ctrl_reg = bitfield_field32_write(
       ctrl_reg, AES_CTRL_SHADOWED_OPERATION_FIELD,
-      params.encrypt ? AES_CTRL_SHADOWED_OPERATION_VALUE_AES_ENC
-                     : AES_CTRL_SHADOWED_OPERATION_VALUE_AES_DEC);
+      encrypt ? AES_CTRL_SHADOWED_OPERATION_VALUE_AES_ENC
+              : AES_CTRL_SHADOWED_OPERATION_VALUE_AES_DEC);
+
+  // Indicate whether the key will be sideloaded.
+  ctrl_reg = bitfield_bit32_write(ctrl_reg, AES_CTRL_SHADOWED_SIDELOAD_BIT,
+                                  key.sideload);
 
-  ctrl_reg = bitfield_field32_write(ctrl_reg, AES_CTRL_SHADOWED_MODE_FIELD,
-                                    params.mode);
+  // Set the mode and the key length.
+  ctrl_reg =
+      bitfield_field32_write(ctrl_reg, AES_CTRL_SHADOWED_MODE_FIELD, key.mode);
   ctrl_reg = bitfield_field32_write(ctrl_reg, AES_CTRL_SHADOWED_KEY_LEN_FIELD,
-                                    params.key_len);
+                                    key.key_len);
 
+  // Never enable manual operation.
   ctrl_reg = bitfield_bit32_write(
       ctrl_reg, AES_CTRL_SHADOWED_MANUAL_OPERATION_BIT, false);
+
+  // Always set the PRNG reseed rate to once per 64 blocks.
+  ctrl_reg =
+      bitfield_field32_write(ctrl_reg, AES_CTRL_SHADOWED_PRNG_RESEED_RATE_FIELD,
+                             AES_CTRL_SHADOWED_PRNG_RESEED_RATE_VALUE_PER_64);
+
   abs_mmio_write32_shadowed(kBase + AES_CTRL_SHADOWED_REG_OFFSET, ctrl_reg);
   err = spin_until(AES_STATUS_IDLE_BIT);
   if (err != kAesOk) {
     return err;
   }
 
-  uint32_t share0 = kBase + AES_KEY_SHARE0_0_REG_OFFSET;
-  uint32_t share1 = kBase + AES_KEY_SHARE1_0_REG_OFFSET;
-
-  for (size_t i = 0; i < key_words; ++i) {
-    abs_mmio_write32(share0 + i * sizeof(uint32_t), params.key[0][i]);
-    abs_mmio_write32(share1 + i * sizeof(uint32_t), params.key[1][i]);
-  }
-  for (size_t i = key_words; i < 8; ++i) {
-    // NOTE: all eight share registers must be written; in the case we don't
-    // have enough key data, we fill it with zeroes.
-    abs_mmio_write32(share0 + i * sizeof(uint32_t), 0);
-    abs_mmio_write32(share1 + i * sizeof(uint32_t), 0);
-  }
-  err = spin_until(AES_STATUS_IDLE_BIT);
+  // Write the key (if it is not sideloaded).
+  err = aes_write_key(key);
   if (err != kAesOk) {
     return err;
   }
 
   // ECB does not need to set an IV, so we're done early.
-  if (params.mode == kAesCipherModeEcb) {
+  if (key.mode == kAesCipherModeEcb) {
     return kAesOk;
   }
 
   uint32_t iv_offset = kBase + AES_IV_0_REG_OFFSET;
-  for (size_t i = 0; i < ARRAYSIZE(params.iv); ++i) {
-    abs_mmio_write32(iv_offset + i * sizeof(uint32_t), params.iv[i]);
+  for (size_t i = 0; i < 4; ++i) {
```

Comment:
nit: Generate a constant for `4` or wrap iv in a struct to be able to use `ARRAYSIZE`